// growtopia.js/dist/structures/Client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Client = void 0;
const events_1 = __importDefault(require("events"));
const native_1 = require("../../native");
class Client extends events_1.default {
    host;
    config;
    constructor(options) {
        super();
        this.config = {
            enet: {
                ip: options?.enet?.ip ?? "127.0.0.1",
                port: options?.enet?.port ?? 17091,
                maxPeers: options?.enet?.maxPeers ?? 1024,
                useNewPacket: {
                    asClient: options?.enet?.useNewPacket?.asClient ?? false
                },
                useNewServerPacket: options?.enet?.useNewServerPacket ?? false,
                channelLimit: options?.enet?.channelLimit ?? 2
            }
        };
        this.host = new native_1.Host(this.config.enet?.ip, this.config.enet?.port, this.config.enet?.maxPeers, this.config.enet?.channelLimit, this.config.enet?.useNewPacket?.asClient, this.config.enet?.useNewServerPacket, options?.enet?.incomingBandwidthLimit ?? null, options?.enet?.outgoingBandwidthLimit ?? null, options?.enet?.enableCompressor ?? null, options?.enet?.enableChecksum ?? null, options?.enet?.seed ?? null);
    }
    on(event, listener) {
        return super.on(event, listener);
    }
    send(netID, channelID, ...data) {
        const packets = data.map((packet) => {
            if (Buffer.isBuffer(packet))
                return packet;
            else {
                switch (packet.constructor.name) {
                    case "TextPacket":
                    case "TankPacket": {
                        return packet.parse();
                    }
                    case "Variant": {
                        return packet.parse().parse();
                    }
                    default: {
                        break;
                    }
                }
            }
        });
        packets
            .filter((packet) => packet !== undefined)
            .forEach((v) => {
            this.host.send(netID, v, channelID);
        });
        return;
    }
    connect(ipAddress, port) {
        return this.host.connect(ipAddress, port);
    }
    listen() {
        try {
            this.host.setEmitter(this.emit.bind(this));
            const acceptPromise = () => new Promise((resolve) => setImmediate(() => resolve(this.host.service())));
            const loop = async () => {
                while (true)
                    await acceptPromise();
            };
            loop();
            this.emit("ready");
        }
        catch {
            this.emit("error", new Error("Failed to initialize ENet server"));
        }
    }
}
exports.Client = Client;
