// growtopia.js/dist/structures/Peer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Peer = void 0;
class Peer {
    client;
    data;
    constructor(client, netID, channelID = 0) {
        this.client = client;
        this.data = {
            netID,
            channelID
        };
        this.client = client;
    }
    get native() {
        return this.client.host.getPeer(this.data.netID);
    }
    /**
     * Sends multiple packets to a single peer.
     * @param data An argument of packets that contains the `parse()` function or just an array of Buffers.
     */
    send(...data) {
        Peer.send(this.client.host, this.data.netID, this.data.channelID, ...data);
    }
    /**
     * Sends multiple packets to a single peer.
     * @param _client The Wrapper client.
     * @param netID The netID of a peer.
     * @param channelID The channelID of a channels.
     * @param data An argument of packets that contains the `parse()` function or just an array of Buffers.
     */
    static send(_client, netID, channelID, ...data) {
        const packets = data.map((packet) => {
            if (Buffer.isBuffer(packet))
                return packet;
            else {
                switch (packet.constructor.name) {
                    case "TextPacket":
                    case "TankPacket": {
                        return packet.parse();
                    }
                    case "Variant": {
                        return packet.parse().parse();
                    }
                    default: {
                        break;
                    }
                }
            }
        });
        packets
            .filter((packet) => packet !== undefined)
            .forEach((v) => {
            _client.send(netID, v, channelID);
        });
        return;
    }
    /**
     * Disconnects the peer.
     * @param type Type of disconnection. Defaults to `later`.
     */
    disconnect(type = "later") {
        const nativePeer = this.native;
        switch (type) {
            case "normal": {
                nativePeer.disconnect(this.client.host, 0);
                break;
            }
            case "later": {
                nativePeer.disconnectLater(this.client.host, 0);
                break;
            }
            case "now": {
                nativePeer.disconnectNow(this.client.host, 0);
                break;
            }
        }
    }
}
exports.Peer = Peer;
