// growtopia.js/dist/Constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PacketTypes = exports.VariantTypes = void 0;
/**
 * Types for each Argument.
 */
var VariantTypes;
(function (VariantTypes) {
    VariantTypes[VariantTypes["NONE"] = 0] = "NONE";
    VariantTypes[VariantTypes["FLOAT_1"] = 1] = "FLOAT_1";
    VariantTypes[VariantTypes["STRING"] = 2] = "STRING";
    VariantTypes[VariantTypes["FLOAT_2"] = 3] = "FLOAT_2";
    VariantTypes[VariantTypes["FLOAT_3"] = 4] = "FLOAT_3";
    VariantTypes[VariantTypes["UNSIGNED_INT"] = 5] = "UNSIGNED_INT";
    VariantTypes[VariantTypes["SIGNED_INT"] = 9] = "SIGNED_INT";
})(VariantTypes || (exports.VariantTypes = VariantTypes = {}));
/**
 * Growtopia Packet Types
 */
var PacketTypes;
(function (PacketTypes) {
    PacketTypes[PacketTypes["UNK"] = 0] = "UNK";
    PacketTypes[PacketTypes["HELLO"] = 1] = "HELLO";
    PacketTypes[PacketTypes["STR"] = 2] = "STR";
    PacketTypes[PacketTypes["ACTION"] = 3] = "ACTION";
    PacketTypes[PacketTypes["TANK"] = 4] = "TANK";
    PacketTypes[PacketTypes["ERROR"] = 5] = "ERROR";
    PacketTypes[PacketTypes["TRACK"] = 6] = "TRACK";
    PacketTypes[PacketTypes["CLIENT_LOG_REQ"] = 7] = "CLIENT_LOG_REQ";
    PacketTypes[PacketTypes["CLIENT_LOG_RES"] = 8] = "CLIENT_LOG_RES";
})(PacketTypes || (exports.PacketTypes = PacketTypes = {}));
