// enet-js/dist/host.js
import koffi from "koffi";
import { ENetEventType } from "./enums.js";
import { enetPacket, enetPeer, enet_host_broadcast, enet_host_connect, enet_host_create, enet_host_destroy, enet_host_flush, enet_host_service, } from "./native/index.js";
import { ipFromLong, ipToLong } from "./util.js";
const broadcast = (host, channelID, packet) => {
    enet_host_broadcast(host.native, channelID, packet.native);
};
const formatAddress = (address) => ({
    host: ipToLong(address.host),
    port: address.port,
});
const formatPeer = (peer) => {
    const peerAttributes = koffi.decode(peer, enetPeer);
    const peerAddress = peerAttributes.address;
    return {
        address: {
            host: ipFromLong(peerAddress.host),
            port: peerAddress.port,
        },
        mtu: peerAttributes.mtu,
        native: peer,
    };
};
const connect = (host, address, channelCount, data) => {
    const peer = enet_host_connect(host.native, formatAddress(address), channelCount, data);
    if (!peer) {
        return null;
    }
    return formatPeer(peer);
};
const create = (address, peerCount, channelLimit, incomingBandwidth, outgoingBandwidth) => {
    const host = enet_host_create(address ? formatAddress(address) : null, peerCount, channelLimit, incomingBandwidth, outgoingBandwidth);
    if (!host) {
        return null;
    }
    return { native: host };
};
const destroy = (host) => {
    enet_host_destroy(host.native);
};
const flush = (host) => {
    enet_host_flush(host.native);
};
const formatPacket = (packet) => {
    const packetAttributes = koffi.decode(packet, enetPacket);
    const dataView = koffi.view(packetAttributes.data, packetAttributes.dataLength);
    return {
        ...packetAttributes,
        data: Buffer.from(dataView),
        native: packet,
    };
};
const formatEvent = (event) => {
    const eventAttributes = event;
    if (eventAttributes.type === ENetEventType.none) {
        return {
            ...eventAttributes,
            native: event,
            packet: null,
            peer: null,
            type: ENetEventType.none,
        };
    }
    if (eventAttributes.type === ENetEventType.receive) {
        return {
            ...eventAttributes,
            native: event,
            packet: formatPacket(eventAttributes.packet),
            peer: formatPeer(eventAttributes.peer),
            type: ENetEventType.receive,
        };
    }
    return {
        ...eventAttributes,
        native: event,
        packet: null,
        peer: formatPeer(eventAttributes.peer),
        type: eventAttributes.type,
    };
};
const service = (host, timeout) => {
    const event = {};
    enet_host_service(host.native, event, timeout);
    return formatEvent(event);
};
export { broadcast, connect, create, destroy, flush, service };
